subsys/net/lib/mqtt/mqtt.c

```diff
@@ -647,3 +647,23 @@ int mqtt_read_publish_payload_blocking(struct mqtt_client *client, void *buffer,
 {
 	return read_publish_payload(client, buffer, length, true);
 }
+
+int mqtt_read_publish_payload_waitall(struct mqtt_client *client, u8_t *buffer,
+			       size_t length)
+{
+	u8_t *end = buffer + length;
+
+	while (buffer < end) {
+		int ret = mqtt_read_publish_payload_blocking(client, buffer,
```

Comment:
So with bad luck, we might hang here forever. Shouldn't there be some kind of timeout because of that?

Reply:
> So with bad luck, we might hang here forever. Shouldn't there be some kind of timeout because of that?

Well, people concerned with non-hanging, timeouts, etc., simply should not use such "convenience" functions at all. They should use elementary functions like mqtt_read_publish_payload() and combine them in a way which makes sense for a particular application. Any convenience functions won't be able to satisfy all possible applications, so as well might not to try.

Reply:
I wouldn't say it'd hang forever - only until TCP connection is open. Same story as with a regular blocking socket.

Reply:
Fair enough, so it's caveat emptor then.